Bio/SeqUtils/__init__.py

```diff
@@ -667,6 +667,40 @@ def calculate(self, sequence):
 
         return exp(cai_value / cai_length)
 
+    def optimize(self, sequence, seq_type="DNA", table=standard_dna_table):
+        """Return a new DNA sequence with preferred codons only."""
+        try:  # If seq record is provided, convert to sequence
+            name = sequence.id
+            sequence = sequence.seq
+        except AttributeError:  # not a  SeqRecord object
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {aminoacid: [] for aminoacid in table.protein_alphabet}
+        for codon, aminoacid in table.forward_table.items():
+            if self[codon] == 1.0:
+                pref_codons[aminoacid] = codon
+        # Create amino acid sequence if DNA was provided
+        if seq_type == "DNA":
+            aa_seq = translate(seq)
+        elif seq_type == "protein":
+            aa_seq = seq
+        else:
+            raise TypeError
+        # Un-translate in loop using only preferred codons
+        optimized = Seq("")
+        for aa in aa_seq:
+            try:
+                pref_codon = pref_codons[aa]
+            except KeyError:  # Keep stop codon from original sequence
+                if aa == "*":
+                    pref_codon = codon
+                    continue
+                else:
+                    print(f"Unrecognized codon: {codon}")
+            optimized += pref_codon
```

Comment:
How long are these strings (and how often might a real use case call this method)?

I've not benchmarked this in years but this used to be true:

Slow:

```
my_string = ""
for letter in some_long_loop:
    my_string += letter
```

Fast:

```
my_letters = []
for letter in some_long_loop:
    my_letters.append(letter)
my_string = "".join(my_letters)
```

Reply:
I imagine this function is only useful for coding sequences, since anything else would not use codons. Coding sequences are usually only a couple thousand base pairs in my experience. They do occasionally get very large though (low millions). I'll benchmark on a very large sequence and get back to you.

Reply:
Timeit results from optimizing titin, one of the largest proteins in the human genome:
Using str: 145 ms ± 18.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Using list: 3.3 ms ± 322 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

Reply:
Thanks - good to know that coding approach is still relevant on modern Python.